forage-android/src/main/java/com/joinforage/forage/android/pos/ForageTerminalSDK.kt

```diff
@@ -300,6 +252,79 @@ class ForageTerminalSDK(
+     * [`Refund`]https://docs.joinforage.app/reference/create-payment-refund) object.
+     * * On failure, the response includes a list of
+     * * [ForageError][com.joinforage.forage.android.network.model.ForageError] objects that you can
+     * * unpack to troubleshoot the issue.
+     *
+     * @param params The [PosRefundPaymentParams] parameters required for refunding a Payment.
+     * @return A [ForageAPIResponse][com.joinforage.forage.android.network.model.ForageApiResponse]
+     * @throws ForageConfigNotSetException If the passed ForagePINEditText instance
+     * hasn't had its ForageConfig set via .setForageConfig().
+     */
+    suspend fun refundPayment(params: PosRefundPaymentParams): ForageApiResponse<String> {
+        val logger = createLogger()
+        val (foragePinEditText, paymentRef, amount, reason) = params
+        val (merchantId, sessionToken) = forageSdk._getForageConfigOrThrow(foragePinEditText)
+
+        val illegalVaultException = isIllegalVaultExceptionOrNull(foragePinEditText, logger)
+        if (illegalVaultException != null) {
+            return illegalVaultException
+        }
+
+        logger
+            .addAttribute("payment_ref", paymentRef)
+            .addAttribute("merchant_ref", merchantId)
+        logger.i(
+            """
+            [POS] Called refundPayment for Payment $paymentRef
+            with amount: $amount
+            for reason: $reason
+            on Terminal: $posTerminalId
+            """.trimIndent()
+        )
+
+        // This block is used for tracking Metrics!
+        // ------------------------------------------------------
+        val measurement = CustomerPerceivedResponseMonitor.newMeasurement(
+            vault = foragePinEditText.getVaultType(),
+            vaultAction = UserAction.REFUND,
+            logger
+        )
+        measurement.start()
+        // ------------------------------------------------------
+
+        val serviceFactory = createServiceFactory(sessionToken, merchantId, logger)
+        val refundService = serviceFactory.createRefundPaymentRepository(foragePinEditText)
+        val refund = refundService.refundPayment(
+            merchantId = merchantId,
+            posTerminalId = posTerminalId,
+            refundParams = params
+        )
+        forageSdk.processApiResponseForMetrics(refund, measurement)
+
+        if (refund is ForageApiResponse.Failure) {
+            logger.e("[POS] refundPayment failed for Payment $paymentRef on Terminal $posTerminalId: ${refund.errors[0]}")
+        }
+
+        return refund
+    }
+
+    private fun isIllegalVaultExceptionOrNull(
+        foragePinEditText: ForagePINEditText,
+        logger: Log
+    ): ForageApiResponse<String>? {
+        if (foragePinEditText.getVaultType() != VaultType.VGS_VAULT_TYPE) {
+            logger.e("[POS] checkBalance failed on Terminal $posTerminalId because the vault type is not VGS")
+            return UnknownErrorApiResponse
```

Comment:
Thoughts on returning back a new developer friendly error to push them to use the POS params?

Reply:
Great suggestion!

See diff here: 7de1243da26a20e9371654523f21bad3f9a70031